### PR TITLE
Revert "Revert "Export useful metrics""

### PR DESCRIPTION
**Pull request by fishbone**
Reverts ray-project/ray#17752

### CODE REVIEW

**Review by rkooo567 — APPROVED** [review]
cc @ashione @lixin-wei  in case you guys would like to review.

File: src/ray/stats/metric.h [hunk 1]
```diff
@@ -191,6 +201,195 @@ struct MetricPoint {
...
+  static constexpr const char *val = "_dist";
+};
+
+template <StatsType T>
+void RegisterView(const std::string &name, const std::string &description,
+                  const std::vector<opencensus::tags::TagKey> &tag_keys,
+                  const std::vector<double> &buckets) {
+  using I = StatsTypeMap<T>;
+  auto view_descriptor = opencensus::stats::ViewDescriptor()
+                             .set_name(name + I::val)
```
Comment (ashione) [inline src/ray/stats/metric.h:253]: I'm not sure why we need add extra suffix for every metric. It's confused if users set one metric name but get other different name from exporter.

**Comment by ashione** [comment]
@iycheng @rkooo567 Could you introduce some doc about this stats implementation?
I got the following points:
* Define metric of stats in marco format
* Register metric with special suffix and recording them in new code path.

**Comment by ashione** [comment]
Please hold on, @jovany-wang might take a look.

**Comment by iycheng** [comment]
@ashione what I did here is trying to enable one-measure-multi-view (not working right now due to backend error) and change the way we define metrics to allow users to use it easier. 

Metrics should be defined in an ad-hoc way so that it can be used right in the place where it's defined and also across functions/classes.

The test failure is because I changed the name. I'll fix it.

File: src/ray/stats/metric.h [hunk 2]
```diff
@@ -191,6 +201,195 @@ struct MetricPoint {
...
+  void Record(double val, std::string tag_val) {
+    RAY_CHECK(tag_keys_.size() == 1);
+    std::unordered_map<std::string, std::string> tags{{tag_keys_[0], std::move(tag_val)}};
+    Record(val, std::move(tags));
+  }
+
+  /// Record a value
+  /// \param val The value to record
+  /// \param tags The tags for this value
+  void Record(double val, std::unordered_map<std::string, std::string> tags) {
```
[the remark this answers is not in the document]
Comment (iycheng) [reply src/ray/stats/metric.h:332]: `&` is not good here as it'll change the value for safety. It's bad to make it change the value from caller.

Another thing here is usually it's called with {} initializer, so it won't have performance issues. 
Or the user need to call it with std::move.

File: src/ray/stats/metric.h [hunk 3]
```diff
@@ -191,6 +201,195 @@ struct MetricPoint {
...
+
+  /// Record a value
+  /// \param val The value to record
+  void Record(double val) { Record(val, std::unordered_map<std::string, std::string>()); }
+
+  /// Record a value
+  /// \param val The value to record
+  /// \param tag_val The tag value. This method will assume we only have one tag for
+  /// this metric.
+  void Record(double val, std::string tag_val) {
```
[the remark this answers is not in the document]
Comment (iycheng) [reply src/ray/stats/metric.h:323]: For customer metrics, it's common, right? This is only for customer tags. You can check the defined metrics in other places.

File: src/ray/stats/metric.h (hunk 1 above)
Comment (iycheng) [reply src/ray/stats/metric.h:253]: It's for one measure multi-view. It's the only way to do this. It explicitly gives the suffix of the type, I think the user knows what's there. Besides, we already add prefixes there.

File: src/ray/stats/metric.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (ashione) [reply src/ray/stats/metric.h:332]: why the tag map parameter of record function will be modified outsie of this scope? we usually mark it in read-only. Did you mean users want to change its tags after recoding invoking?

**Comment by ashione** [comment]
> @ashione what I did here is trying to enable one-measure-multi-view (not working right now due to backend error) and change the way we define metrics to allow users to use it easier.
> 
> Metrics should be defined in an ad-hoc way so that it can be used right in the place where it's defined and also across functions/classes.
> 
> The test failure is because I changed the name. I'll fix it.

Is it good solution to bind a measure for multiple views? Addtionally, there will be heavy overhead of count and histogram.
In my thoughts, stats had been designed in low level library or api, so c++/java/python etc. workers can build an enhancement upper metric system to user level by stats lib via cython or jni.

**Comment by jovany-wang** [comment]
Hi @iycheng @rkooo567  The micros look very good and easy to use. 

For multiple views, I didn't found where we use multiple views on one measure in this PR. Could you tell us more about in which case we need this?
Another question is what if  user does specify one measure on both Gauge and Histogram?

File: src/ray/stats/metric.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (iycheng) [reply src/ray/stats/metric.h:332]: No that's for safety. Please check the implementation of this one and I have a comment for this. Ideally, we don't need it, but sometimes it happens, and we do have some memory corruption in caller_name_ right now,.

**Comment by iycheng** [comment]
> > @ashione what I did here is trying to enable one-measure-multi-view (not working right now due to backend error) and change the way we define metrics to allow users to use it easier.
> > Metrics should be defined in an ad-hoc way so that it can be used right in the place where it's defined and also across functions/classes.
> > The test failure is because I changed the name. I'll fix it.
> 
> Is it good solution to bind a measure for multiple views? Addtionally, there will be heavy overhead of count and histogram.
> In my thoughts, stats had been designed in low level library or api, so c++/java/python etc. workers can build an enhancement upper metric system to user level by stats lib via cython or jni.

No the overload should be lower. Measure is about the thing you are counting. View is the thing you aggregate the results. Ideally, you measure once and have multi view for this. This saves time. Measure is the place where time spending on collector side.

Think about this, COUNT/SUM -> you only send one value, and you'll get both.

**Comment by iycheng** [comment]
> Hi @iycheng @rkooo567 The micros look very good and easy to use.
> 
> For multiple views, I didn't found where we use multiple views on one measure in this PR. Could you tell us more about in which case we need this?
> Another question is what if user does specify one measure on both Gauge and Histogram?

It's supposed to be done by opencensus automatically. (https://opencensus.io/quickstart/cpp/metrics/) Check the measure  kLatencyMeasureName. cc @ashione 

But right now, there is a bug in the backend preventing us from doing this. So even you set it up, it won't work. But later we fix that bug and it should just work.

**Comment by rkooo567** [comment]
Lmk when I can merge this. For https://github.com/ray-project/ray/pull/17755#issuecomment-898193269, I think we can probably have a simple example in the docstring.

**Comment by iycheng** [comment]
@rkooo567  I think I've put a docstring in `DEFINE_stats`. Do you have other suggestions about this?

**Comment by jovany-wang** [comment]
I see. Sound good if it can be done automatically on client side(opencensus side).

**Comment by iycheng** [comment]
@jovany-wang I don't think Prometheus will need change, if you check this (https://opencensus.io/quickstart/cpp/metrics/), it's actually using Prometheus (check the end of the charts).

Another thing is that, we actually use the measure name right now, so the suffix is not added until we fixed it in the agent.

**Comment by ashione** [comment]
> It's supposed to be done by opencensus automatically. (https://opencensus.io/quickstart/cpp/metrics/) Check the measure kLatencyMeasureName. cc @ashione



> > > @ashione what I did here is trying to enable one-measure-multi-view (not working right now due to backend error) and change the way we define metrics to allow users to use it easier.
> > > Metrics should be defined in an ad-hoc way so that it can be used right in the place where it's defined and also across functions/classes.
> > > The test failure is because I changed the name. I'll fix it.
> > 
> > 
> > Is it good solution to bind a measure for multiple views? Addtionally, there will be heavy overhead of count and histogram.
> > In my thoughts, stats had been designed in low level library or api, so c++/java/python etc. workers can build an enhancement upper metric system to user level by stats lib via cython or jni.
> 
> No the overload should be lower. Measure is about the thing you are counting. View is the thing you aggregate the results. Ideally, you measure once and have multi view for this. This saves time. Measure is the place where time spending on collector side.
> 
> Think about this, COUNT/SUM -> you only send one value, and you'll get both.

Do we need one measure V.S. multiple views in our production usages, actually? Could you show me some examples like this? Data aggregation function will be provided by modern stats platform (opentsdb/promehues or others).

**Comment by iycheng** [comment]
> > It's supposed to be done by opencensus automatically. (https://opencensus.io/quickstart/cpp/metrics/) Check the measure kLatencyMeasureName. cc @ashione
> 
> > > > @ashione what I did here is trying to enable one-measure-multi-view (not working right now due to backend error) and change the way we define metrics to allow users to use it easier.
> > > > Metrics should be defined in an ad-hoc way so that it can be used right in the place where it's defined and also across functions/classes.
> > > > The test failure is because I changed the name. I'll fix it.
> > > 
> > > 
> > > Is it good solution to bind a measure for multiple views? Addtionally, there will be heavy overhead of count and histogram.
> > > In my thoughts, stats had been designed in low level library or api, so c++/java/python etc. workers can build an enhancement upper metric system to user level by stats lib via cython or jni.
> > 
> > 
> > No the overload should be lower. Measure is about the thing you are counting. View is the thing you aggregate the results. Ideally, you measure once and have multi view for this. This saves time. Measure is the place where time spending on collector side.
> > Think about this, COUNT/SUM -> you only send one value, and you'll get both.
> 
> Do we need one measure V.S. multiple views in our production usages, actually? Could you show me some examples like this? Data aggregation function will be provided by modern stats platform (opentsdb/promehues or others).

@ashione We definitely need this one! We'll want to check request count/request latency and this can be done with one measure. We'll want to check the count and sum of bytes and even the p95 of objects added, and this can be done with one measure. Why duplicate counting them? Why sending two RPCs for the same metrics? Although the current job agent doesn't support it, it doesn't mean that it's not needed. Could you give me some highlights of why you think it's not needed? 
 
It's supported in opencensus as I mentioned before. I haven't checked opentelemetry yet, but they also have concepts about measure/metrics.

I don't think we'll use Prometheus directly, and all exporters will do their own job for pre-aggregation first and send it to Prometheus and the backend will do aggregation again.

**Comment by rkooo567** [comment]
Looks like the only discussion point is the multiple view for the same measure, and multiple view is anyway not enabled now, so I think it is fine to merge it. We can have further discussion in the future PRs that enables it.

**Pull request merged**